website/static/js/setPassword.js

```diff
@@ -0,0 +1,234 @@
+'use strict';
+
+var ko = require('knockout');
+require('knockout.validation');
+var $ = require('jquery');
+var zxcvbn = require('zxcvbn');
+
+var $osf = require('./osfHelpers');
+
+
+ko.validation.rules.complexity = {
+    validator: function (val, minimumComplexity) {
+        return zxcvbn(val).score >= minimumComplexity;
+    },
+    message: 'Please enter a more complex password.'
+};
+
+ko.validation.registerExtenders();
+
+
+// Accepts a few different types of arguments, depending on
+// the desired fields and type of password reset.
+// types:
+//    - signup: required fields for Name, email, email confirmation, password, and password strength
+//    - reset: fields for password, strength, and password confirmation (for when you don't know your old password)
+//    - change: change your password when you know your old one
+var ViewModel = function(passwordViewType, submitUrl, campaign, redirectUrl) {
+
+    var self = this;
+
+    self.typedPassword = ko.observable('');
+
+    self.passwordInfo = ko.computed(function() {
+        if (self.typedPassword()) {
+            return zxcvbn(self.typedPassword().slice(0, 100));
+        }
+    });
+
+    self.passwordFeedback = ko.computed(function () {
+        if (self.typedPassword()) {
+            return self.passwordInfo().feedback.warning;
+        }
+    });
+
+    self.passwordComplexity = ko.pureComputed(function() {
+        if (self.typedPassword()) {
+            return self.passwordInfo().score;
+        }
+    });
+
+    self.passwordComplexityBar = ko.computed(function() {
+        return $osf.valueProgressBar(self.passwordComplexity());
+    });
+
+    self.password = ko.observable('').extend({
+        required: true,
+        minLength: 6,
+        maxLength: 256,
+        complexity: 2,
+    });
+
+    self.campaign = ko.observable(campaign);
+
+    // Preserve object of validated fields for use in `submit`
+    var validatedFields = {
+        password: self.password
+    };
+
+    if (passwordViewType === 'reset' || passwordViewType === 'change') {
```

Comment:
After further investigation, approach 1) is totally possible! KO allows you to define custom "extenders": http://knockoutjs.com/documentation/extenders.html .

But again, this doesn't need to happen before this is merged. I've created a follow-up ticket for this: https://openscience.atlassian.net/browse/OSF-6514